### PR TITLE
Added method to check for escrow duration, fixes #874

### LINKED ISSUE #874: Escrow End Date
Bot is unable to get this param before or after the trade.
For example, if bot will see that items in offer will be holded - bot can decline offer.

### PR DESCRIPTION
**Pull request by waylaidwanderer**
/u/starsintheeyes on /r/SteamBot was kind enough to write a function to check for escrow duration. I've modified his code to work with SteamBot.

~~I haven't tested it personally, but it looks good to me.~~
Edit: confimed working.

Fixes #874

### CODE REVIEW

**Comment by TheAnthonyNL** [comment]
But what about escrow state on receiving trade offers? Since this is only for sending.

**Comment by waylaidwanderer** [comment]
Check if `Offer.EscrowEndDate` is 0 or not.

**Comment by scholtzm** [comment]
Going to be set after you accept the offer. You need to check the escrow duration before doing anything.

**Comment by waylaidwanderer** [comment]
@scholtzm So if I'm wrong, how do you do that for receiving trade offers? All my bots only send trade offers, so I have no clue about this end of things.

**Comment by TheAnthonyNL** [comment]
@waylaidwanderer how does your bot only send trade offers? Doesn't it get incoming trade offers at all? 

One likely solution is to send trade offer to that person before accept/decline the trade offer but that would be little bit overkill...

**Comment by scholtzm** [comment]
@waylaidwanderer The code is pretty much the same but the URL you use is `http://steamcommunity.com/tradeoffer/<tradeofferid>`. Patching the current PR should be simple enough.

File: SteamBot/Bot.cs [hunk 1]
```diff
@@ -1083,6 +1084,55 @@ public void AcceptAllMobileTradeConfirmations()
...
         }
+        
+        /// <summary>
+        /// Get duration of escrow in days. Call this before sending a trade offer.
+        /// Credit to: https://www.reddit.com/r/SteamBot/comments/3w8j7c/code_getescrowduration_for_c/
+        /// </summary>
+        /// <param name="partnerId">Partner ID of user you want to send a trade offer to</param>
+        /// <param name="token">User's trade token. Can be an empty string if user is on bot's friends list.</param>
+        /// <returns>TradeOfferEscrowDuration</returns>
+        public TradeOfferEscrowDuration GetEscrowPartnerDuration(int partnerId, string token)
```
Comment (scholtzm) [inline SteamBot/Bot.cs]: Probably better to use `SteamID` as the first argument. Can be easily "converted" to AccountID (= PartnerID).

**Comment by waylaidwanderer** [comment]
@TheAnthonyNL My bot automatically rejects all incoming trade offers. The way it works is the user selects which items they want to buy/sell via the site, and then the bot sends a trade offer with those items.

@scholtzm I'll make the changes and function you suggested.

**Comment by TheAnthonyNL** [comment]
@waylaidwanderer  gotcha :D.

**Comment by waylaidwanderer** [comment]
@scholtzm Dumb question but should tradeofferid be uint or ulong?

**Comment by scholtzm** [comment]
You receive `TradeOffer` object in `OnNewTradeOffer` method, and the `TradeOffer` class uses `string` apparently: https://github.com/Jessecar96/SteamBot/blob/master/SteamTrade/TradeOffer/TradeOffer.cs#L18

I would make it a `string` in that case. Using numbers would require parsing.

**Comment by waylaidwanderer** [comment]
@scholtzm Alright, that would probably be the best way to do it. In my SteamTradeOffersBot fork I have them set as ulongs.

**Comment by waylaidwanderer** [comment]
@scholtzm Have a look: https://github.com/waylaidwanderer/SteamTradeOffersBot/commit/8f1a4e2d69f9ea3816a7c9083e34761283cc3810

File: SteamBot/Bot.cs [hunk 2]
```diff
@@ -1083,6 +1084,74 @@ public void AcceptAllMobileTradeConfirmations()
                 }
             }                        
         }
+        
+        /// <summary>
+        /// Get duration of escrow in days. Call this before sending a trade offer.
+        /// Credit to: https://www.reddit.com/r/SteamBot/comments/3w8j7c/code_getescrowduration_for_c/
+        /// </summary>
+        /// <param name="partnerId">Steam ID of user you want to send a trade offer to</param>
```
Comment (scholtzm) [inline SteamBot/Bot.cs]: old param name

File: SteamBot/Bot.cs [hunk 3]
```diff
@@ -1083,6 +1084,74 @@ public void AcceptAllMobileTradeConfirmations()
...
+            var resp = SteamWeb.Fetch(url, "GET", null, false);
+            if (string.IsNullOrWhiteSpace(resp))
+            {
+                throw new NullReferenceException(nameof(resp));
+            }
+
+            return ParseEscrowResponse(resp);
+        }
+
+        TradeOfferEscrowDuration ParseEscrowResponse(string resp)
```
Comment (scholtzm) [inline SteamBot/Bot.cs]: this should be a private method for sure

File: SteamBot/Bot.cs (hunk 2 above)
Comment (waylaidwanderer) [inline SteamBot/Bot.cs]: whoops

File: SteamBot/Bot.cs (hunk 3 above)
Comment (waylaidwanderer) [inline SteamBot/Bot.cs]: Good point

**Comment by waylaidwanderer** [comment]
I don't know about you guys, but everything looks good to me.

**Comment by scholtzm** [comment]
Looks good to me as well.

cc @BlueRaja

**Comment by TheAnthonyNL** [comment]
Found one annoying thing. If people want to use this code they are forced to use VS 2015 :-1:  http://stackoverflow.com/questions/32943593/the-name-nameof-does-not-exist-in-the-current-context is the error i get in Visual Studio 2013.

**Comment by scholtzm** [comment]
It's not really related to VS but rather the compiler. Either way, that's a good catch.

**Comment by TheAnthonyNL** [comment]
Anyway i tested this part of the code and it works really good.

http://puu.sh/lRpnh.png

**Comment by waylaidwanderer** [comment]
@TheAnthonyNL Just pushed a commit to fix it.

**Comment by waylaidwanderer** [comment]
Uh I tried rebasing some stuff and broke it lol. I'll delete this comment once I fix it.

**Comment by waylaidwanderer** [comment]
I just rebased everything into 1 commit.

**Comment by AleksaSavic3** [comment]
Great work.

File: SteamBot/Bot.cs [hunk 4]
```diff
@@ -1084,6 +1085,74 @@ public void AcceptAllMobileTradeConfirmations()
...
+        }
+
+        /// <summary>
+        /// Get duration of escrow in days. Call this after receiving a trade offer.
+        /// </summary>
+        /// <param name="tradeOfferId">The ID of the trade offer</param>
+        /// <returns>TradeOfferEscrowDuration</returns>
+        public TradeOfferEscrowDuration GetEscrowDuration(string tradeOfferId)
+        {
+            var url = "http://steamcommunity.com/tradeoffer/" + tradeOfferId;
```
Comment (BlueRaja) [inline SteamBot/Bot.cs]: There's no way to get this from the API? :\

Comment (TheAnthonyNL) [inline SteamBot/Bot.cs]: If there was, there would be a solution for it.

Anyway tjhe problem mainly is that  "escrow_end_date" and k_ETradeOfferStateInEscrow only get set after trade accepted or declined :S.

**Pull request merged**